src/net/atlas/mod.rs

```diff
@@ -0,0 +1,137 @@
+pub mod db;
+pub mod download;
+pub mod onchain;
+
+pub use self::db::AtlasDB;
+pub use self::download::AttachmentsDownloader;
+pub use self::onchain::OnchainInventoryLookup;
+
+use chainstate::stacks::boot::boot_code_id;
+use chainstate::stacks::{StacksBlockHeader, StacksBlockId};
+
+use chainstate::burn::db::sortdb::SortitionDB;
+use chainstate::burn::{BlockHeaderHash, ConsensusHash};
+use net::StacksMessageCodec;
+use util::hash::{to_hex, Hash160, MerkleHashFunc};
+use vm::types::{QualifiedContractIdentifier, SequenceData, TupleData, Value};
+
+use std::collections::{HashMap, HashSet};
+use std::convert::TryFrom;
+use std::hash::{Hash, Hasher};
+
+pub const BNS_NAMESPACE_MIN_LEN: usize = 1;
+pub const BNS_NAMESPACE_MAX_LEN: usize = 19;
+pub const BNS_NAME_MIN_LEN: usize = 1;
+pub const BNS_NAME_MAX_LEN: usize = 16;
+
+lazy_static! {
+    pub static ref BNS_NAME_REGEX: String = format!(
+        r#"([a-z0-9]|[-_]){{{},{}}}\.([a-z0-9]|[-_]){{{},{}}}(\.([a-z0-9]|[-_]){{{},{}}})?"#,
+        BNS_NAMESPACE_MIN_LEN, BNS_NAMESPACE_MAX_LEN, BNS_NAME_MIN_LEN, BNS_NAME_MAX_LEN, 1, 128
+    );
+}
+
+pub struct AtlasConfig {
+    pub contracts: HashSet<QualifiedContractIdentifier>,
+    pub attachments_max_size: u32,
+}
+
+impl AtlasConfig {
+    pub fn default() -> AtlasConfig {
+        let mut contracts = HashSet::new();
+        contracts.insert(boot_code_id("bns"));
+        AtlasConfig {
+            contracts,
+            attachments_max_size: 1_048_576,
+        }
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, Eq, PartialEq, Hash)]
+pub struct Attachment {
+    pub hash: Hash160,
```

Comment:
I'd recommend removing the `hash` value, and instead making it a derived value via a `hash()` method.  Then, you won't have to worry about `hash` and `content` becoming inconsistent somehow.

Reply:
Addressed with https://github.com/blockstack/stacks-blockchain/pull/2013/commits/54e632f3f10aad4e631f5a2fe35575228e452f46